vertical-pod-autoscaler/pkg/admission-controller/logic/server_test.go

```diff
@@ -0,0 +1,284 @@
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			nil,
+			fmt.Errorf("unexpected end of JSON input"),
+		},
+		{
+			"invalid pod",
+			[]byte("{}"),
+			"default",
+			fmt.Errorf("bad pod"),
+			[]ContainerResources{},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			nil,
+			fmt.Errorf("bad pod"),
+		},
+		{
+			"new cpu recommendation",
+			[]byte(
+				`{
+					"spec": {
+						"containers": [{}]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("1"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourcesPatch(0),
+				addRequestsPatch(0),
+				addResourceRequestPatch(0, cpu, "1"),
+				addAnnotationRequest([][]string{{cpu}}),
+			},
+			nil,
+		},
+		{
+			"replacement cpu recommendation",
+			[]byte(
+				`{
+					"spec": {
+						"containers": [
+							{
+								"resources": {
+									"requests": {
+										"cpu": "0"
+									}
+								}
+							}
+						]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("1"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourceRequestPatch(0, cpu, "1"),
+				addAnnotationRequest([][]string{{cpu}}),
+			},
+			nil,
+		},
+		{
+			"two replacement resources",
+			[]byte(
+				`{
+					"spec": {
+						"containers": [
+							{
+								"resources": {
+									"requests": {
+										"cpu": "0"
+									}
+								}
+							}
+						]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu:        resource.MustParse("1"),
+						unobtanium: resource.MustParse("2"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourceRequestPatch(0, cpu, "1"),
+				addResourceRequestPatch(0, unobtanium, "2"),
+				addAnnotationRequest([][]string{{cpu, unobtanium}}),
+			},
+			nil,
+		},
+		{
+			"two containers",
+			[]byte(
+				`{
+					"spec": {
+						"containers": [
+							{
+								"resources": {
+									"requests": {
+										"cpu": "0"
+									}
+								}
+							},
+							{}
+						]
+					}
+				}`),
+			"default",
+			nil,
+			[]ContainerResources{
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("1"),
+					},
+				},
+				{
+					apiv1.ResourceList{
+						cpu: resource.MustParse("2"),
+					},
+				},
+			},
+			vpa_api_util.ContainerToAnnotationsMap{},
+			"name",
+			nil,
+			[]patchRecord{
+				addResourceRequestPatch(0, cpu, "1"),
+				addResourcesPatch(1),
+				addRequestsPatch(1),
+				addResourceRequestPatch(1, cpu, "2"),
+				addAnnotationRequest([][]string{{cpu}, {cpu}}),
+			},
+			nil,
+		},
+	}
+	for _, tc := range tests {
+		fppp := fakePodPreProcessor{e: tc.preProcessorError}
```

Comment:
You can use t.Run(caseName, func(t *testing.T){}) for nice output of failed tests. Example: https://github.com/kubernetes/autoscaler/blob/b1995a3477cf7578b795c05c96c11a24bdd19388/vertical-pod-autoscaler/pkg/admission-controller/logic/recommendation_provider_test.go#L163

Reply:
Done.